numpy/_core/src/multiarray/einsum_sumprod.cpp

```diff
@@ -0,0 +1,2181 @@
+
+        const SimdReg a01 = SimdType::npyv_add(a0, a1);
+        const SimdReg a23 = SimdType::npyv_add(a2, a3);
+        const SimdReg a0123 = SimdType::npyv_add(a01, a23);
+        v_accum = SimdType::npyv_add(a0123, v_accum);
+    }
+
+    for (; count > 0; count -= vstep, data += vstep) {
+        SimdReg a = SimdType::load_tillz(data, count);
+        v_accum = SimdType::npyv_add(a, v_accum);
+    }
+    T accum = SimdType::npyv_sum(v_accum);
+    SimdType::cleanup();
+    return accum;
+}
+
+template <typename T, typename AccumType>
+static inline NPY_GCC_OPT_3 AccumType
+elementwise_sum_of_arr(const T *data, npy_intp count)
+{
+    AccumType accum = 0;
+
+#ifndef NPY_DISABLE_OPTIMIZATION
+    for (; count > 4; count -= 4, data += 4) {
+        const AccumType a01 = from<T, AccumType>(*data) + from<T, AccumType>(data[1]);
+        const AccumType a23 = from<T, AccumType>(data[2]) + from<T, AccumType>(data[3]);
+        accum += a01 + a23;
+    }
+#endif  // NPY_DISABLE_OPTIMIZATION
+
+    for (; count > 0; --count, ++data) {
+        accum += from<T, AccumType>(*data);
+    }
+    return accum;
+}
+
+template <typename T, typename AccumType>
+struct SumOfArr {
+    static inline NPY_GCC_OPT_3 AccumType eval(T *data, npy_intp count) noexcept
+    {
+        return elementwise_sum_of_arr<T, AccumType>(data, count);
+    }
+};
+
+/* Template where (npy_float, npy_float) will allow the SIMD
+ * capable version.*/
+template <>
+struct SumOfArr<npy_float, npy_float> {
+    static inline NPY_GCC_OPT_3 npy_float eval(npy_float *data, npy_intp count) noexcept
+    {
+#if NPY_SIMD_F32
+        using SimdType = typename SumSIMD<npy_float, npy_float>::SimdType;
+        return floating_point_sum_of_arr<SimdType>(data, count);
+#else
+        return elementwise_sum_of_arr<npy_float, npy_float>(data, count);
+#endif
+    }
+};
+
+/* Template where (npy_double, npy_double) will allow the SIMD
+ * capable version.*/
+template <>
+struct SumOfArr<npy_double, npy_double> {
+    static inline NPY_GCC_OPT_3 npy_double eval(npy_double *data,
+                                                npy_intp count) noexcept
+    {
+#if NPY_SIMD_F64
+        using SimdType = typename SumSIMD<npy_double, npy_double>::SimdType;
+        return floating_point_sum_of_arr<SimdType>(data, count);
+#else
+        return elementwise_sum_of_arr<npy_double, npy_double>(data, count);
+#endif
+    }
+};
+
+template <typename T, typename AccumType>
+static inline NPY_GCC_OPT_3 AccumType
+sum_of_arr(T *data, npy_intp count) noexcept
+{
+    return SumOfArr<T, AccumType>::eval(data, count);
+}
+
+template <typename SimdType>
+static inline NPY_GCC_OPT_3 void
+floating_point_sum_of_products_muladd(const typename SimdType::T *data,
+                                      typename SimdType::T *data_out,
+                                      typename SimdType::AccumType scalar,
+                                      npy_intp count)
```

Comment:
```suggestion
template <typename Lane, typename LaneAcc>
static inline NPY_GCC_OPT_3 void
floating_point_sum_of_products_muladd(const Lane *data,
                                      Lane *data_out,
                                      LaneAcc scalar,
                                      SSize count)
```
Just deduce the scalar type directly; regarding `npy_intp`, you can use `SSize`, which maps to `Py_ssize_t` or `ptrdiff_t`.